### ISSUE TITLE
`map_dict` with `sink_parquet` panics

### ISSUE THREAD

**Issue by TylerGrantSmith** [pola-rs/polars#8553]
### Polars version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Issue description

I get a panic exception about unreachable code when using a `map_dict` in conjunction with `sink_parquet`
```
thread '<unnamed>' panicked at 'internal error: entered unreachable code', D:\a\polars\polars\polars\polars-core\src\datatypes\dtype.rs:250:24
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Traceback (most recent call last):
  File "c:\Users\Tyler\Python\oddsfellow\test.py", line 4, in <module>
    pl.LazyFrame().with_columns(pl.lit(1).alias("x").map_dict({1:2})).sink_parquet("out.parquet")
  File "C:\Users\Tyler\AppData\Local\Programs\Python\Python311\Lib\site-packages\polars\lazyframe\frame.py", line 1698, in sink_parquet
    return lf.sink_parquet(
           ^^^^^^^^^^^^^^^^
pyo3_runtime.PanicException: internal error: entered unreachable code
```
### Reproducible example

```python
import polars as pl 
pl.LazyFrame().with_columns(pl.lit(1).alias("x").map_dict({1:2})).sink_parquet("out.parquet")
```

### Expected behavior

I don't know if it is expected for `map_dict` to be supported in streaming mode, but in lieu of that it should give the usual error when a query is incompatible with `sink_parquet`.

### Installed versions

<details>

```
---Version info---
Polars: 0.17.9
Index type: UInt32
Platform: Windows-10-10.0.19045-SP0
Python: 3.11.1 (tags/v3.11.1:a7a450f, Dec  6 2022, 19:58:39) [MSC v.1934 64 bit (AMD64)]
---Optional dependencies---
numpy: 1.24.3
pandas: 2.0.1
pyarrow: 11.0.0
connectorx: <not installed>
deltalake: <not installed>
fsspec: <not installed>
matplotlib: <not installed>
xlsx2csv: <not installed>
xlsxwriter: <not installed>```

</details>

**Comment by ritchie46** [comment]
@ghuls this is due to the `dtype` becoming `Unknown` somewhere. Setting the `return_dtype` does not work, is there any operation where we must set the dtype as well?

**Comment by ghuls** [comment]
@ritchie46 The dtype unknown thing happens in more cases. It would be nice if lazyframe could fill in the schema when `return_dtype` is explicitly set in `map_dict`: https://github.com/pola-rs/polars/issues/7700

**Comment by ghuls** [comment]
@ritchie46 Now to avoid the `Unknown` dtype, you need to add a cast. When the user sets the `return_dtype`, a cast shouldn't be needed for it to know the schema.

```python
In [15]: df.with_columns(pl.lit(1).alias("x").map_dict({1:2})).schema
Out[15]: {'x': Unknown}

In [16]: df.with_columns(pl.lit(1).cast(pl.Int32).alias("x")).with_columns(pl.col("x").map_dict({1:2}, return_dtype=pl.Int32).cast(pl.Int32)).schema
Out[16]: {'x': Int32}

In [17]: df.with_columns(pl.lit(1).cast(pl.Int32).alias("x")).with_columns(pl.col("x").map_dict({1:2}, return_dtype=pl.Int32).cast(pl.Int32)).sink_parquet("out.parquet")
```

**Comment by Guillermogsjc** [comment]
hello, is there any roadmap to fix this? on 0.19.10 still panics, `sink_parquet` is a really nice feature .

As far as I can try based on @ghuls comment, it is only needed to perform `.cast` **after** the `.map_dict`.

If it is not going to be updated at internals, could be this previous tip be added into `sink_parquet` documentation as a WARNING?

regards

**Comment by stinodego** [comment]
This now raises an `InvalidOperationError` due to `replace` (the new name for `map_dict`) not yet being available in the streaming engine.
